slatedb/src/db_snapshot.rs

```diff
@@ -0,0 +1,618 @@
+    pub async fn scan_with_options<K, T>(
+        &self,
+        range: T,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, crate::Error>
+    where
+        K: AsRef<[u8]> + Send,
+        T: RangeBounds<K> + Send,
+    {
+        // TODO: this range conversion logic can be extract to an util
+        let start = range
+            .start_bound()
+            .map(|b| Bytes::copy_from_slice(b.as_ref()));
+        let end = range
+            .end_bound()
+            .map(|b| Bytes::copy_from_slice(b.as_ref()));
+        let range = (start, end);
+        self.db_inner.check_error()?;
+        let db_state = self.db_inner.state.read().view();
+        self.db_inner
+            .reader
+            .scan_with_options(
+                BytesRange::from(range),
+                options,
+                &db_state,
+                Some(self.txn_state.seq),
+            )
+            .await
+            .map_err(Into::into)
+    }
+}
+
+#[async_trait::async_trait]
+impl DbRead for DbSnapshot {
+    async fn get_with_options<K: AsRef<[u8]> + Send>(
+        &self,
+        key: K,
+        options: &ReadOptions,
+    ) -> Result<Option<Bytes>, crate::Error> {
+        self.get_with_options(key, options).await
+    }
+
+    async fn scan_with_options<K, T>(
+        &self,
+        range: T,
+        options: &ScanOptions,
+    ) -> Result<DbIterator, crate::Error>
+    where
+        K: AsRef<[u8]> + Send,
+        T: RangeBounds<K> + Send,
+    {
+        self.scan_with_options(range, options).await
+    }
+}
+
+/// Unregister from transaction manager when dropped.
+impl Drop for DbSnapshot {
+    fn drop(&mut self) {
+        self.txn_manager.remove_txn(self.txn_state.as_ref());
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rstest::rstest;
+
+    use super::*;
+    use crate::config::{CompactorOptions, Settings};
+    use crate::object_store::memory::InMemory;
+    use crate::object_store::ObjectStore;
+    use crate::{Db, Error};
+    use std::future::Future;
+    use std::pin::Pin;
+    use std::time::Duration;
+
+    type SnapshotTestCaseSetupFunc =
+        fn(&Db) -> Pin<Box<dyn Future<Output = Result<Arc<DbSnapshot>, Error>> + Send + '_>>;
+
+    struct SnapshotGetTestCase {
+        name: &'static str,
+        setup: SnapshotTestCaseSetupFunc,
+        expected_snapshot_results: Vec<(&'static str, Option<&'static str>)>,
+        expected_db_results: Option<Vec<(&'static str, Option<&'static str>)>>,
+    }
+
+    struct SnapshotScanTestCase {
+        name: &'static str,
+        setup: SnapshotTestCaseSetupFunc,
+        scan_start_key: &'static str,
+        expected_snapshot_results: Vec<(&'static str, &'static str)>, // (key, value) pairs
+        expected_db_results: Option<Vec<(&'static str, &'static str)>>,
+    }
+
+    async fn create_test_db() -> Db {
+        let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
```

Comment:
You don't need to declare the type here, Rust's type inference should take care of it:
```suggestion
        let object_store = Arc::new(InMemory::new());
```

Reply:
thank you for this!